Replace a number of FxHashMaps/Sets with stable-iteration-order alternatives

michaelwoerister opened the pull request:
This PR replaces almost all of the remaining `FxHashMap`s in query results with either `FxIndexMap` or `UnordMap`. The only case that is missing is the `EffectiveVisibilities` struct which turned out to not be straightforward to transform. Once that is done too, we can remove the `HashStable` implementation from `HashMap`.

The first commit adds the `StableCompare` trait which is a companion trait to `StableOrd`. Some types like `Symbol` can be compared in a cross-session stable way, but their `Ord` implementation is not stable. In such cases, a `StableCompare` implementation can be provided to offer a lightweight way for stable sorting. The more heavyweight option is to sort via `ToStableHashKey`, but then sorting needs to have access to a stable hashing context and `ToStableHashKey` can also be expensive as in the case of `Symbol` where it has to allocate a `String`.

The rest of the commits are rather mechanical and don't overlap, so they are best reviewed individually.

Part of [MCP 533](https://github.com/rust-lang/compiler-team/issues/533).

michaelwoerister committed:
Split StableCompare trait out of StableOrd trait.

StableCompare is a companion trait to `StableOrd`. Some types like `Symbol` can be compared in a cross-session stable way, but their `Ord` implementation is not stable. In such cases, a `StableOrd` implementation can be provided to offer a lightweight way for stable sorting. (The more heavyweight option is to sort via `ToStableHashKey`, but then sorting needs to have access to a stable hashing context and `ToStableHashKey` can also be expensive as in the case of `Symbol` where it has to allocate a `String`.)

michaelwoerister committed:
Replace a number of FxHashMaps/Sets with stable-iteration-order alternatives.

michaelwoerister committed:
Make iteration order of crate_inherent_impls query result stable.

michaelwoerister committed:
Make iteration order of upstream_monomorphizations query stable

michaelwoerister committed:
Make iteration order of inferred_outlives_crate query stable

michaelwoerister committed:
Make iteration order of wasm_import_module_map query stable

michaelwoerister committed:
Make iteration order of stability_index query stable

michaelwoerister committed:
Make iteration order of named_variable_map, late_bound_vars_map, and resolve_bound_vars queries stable

michaelwoerister committed:
Make iteration order of region_scope_tree query stable

michaelwoerister committed:
Make iteration order of trimmed_def_paths query stable

michaelwoerister committed:
Make iteration order of supported_target_features query stable

michaelwoerister committed:
Make iteration order of collect_return_position_impl_trait_in_trait_tys query stable

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 178b8131d71317c570ed96ddfa3b59efdade066d with merge 28f9060f96b97141dab0e267d633b408d5ec20c1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"178b8131d71317c570ed96ddfa3b59efdade066d","merge_sha":"28f9060f96b97141dab0e267d633b408d5ec20c1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7290519554/job/19870240234)
Build commit: 28f9060f96b97141dab0e267d633b408d5ec20c1 (`28f9060f96b97141dab0e267d633b408d5ec20c1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7290519554/job/19870240234"},"merge_sha":"28f9060f96b97141dab0e267d633b408d5ec20c1"} -->

rust-timer commented:
Queued 28f9060f96b97141dab0e267d633b408d5ec20c1 with parent 767453eb7ca188e991ac5568c17b984dd4893e77, future [comparison URL](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (28f9060f96b97141dab0e267d633b408d5ec20c1): [comparison URL](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.1%, 0.6%] | 33    |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.1%, 0.9%] | 39    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.1%, 0.6%] | 33    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [0.7%, 1.9%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 1.6%  | [0.8%, 2.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.7%, -3.2%] | 2     |
| All ❌✅ (primary)                 | 1.3%  | [0.7%, 1.9%]   | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1% | [3.1%, 3.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=767453eb7ca188e991ac5568c17b984dd4893e77&end=28f9060f96b97141dab0e267d633b408d5ec20c1&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

**Bootstrap**: 673.174s -> 673.426s (0.04%)
**Artifact size**: 312.82 MiB -> 312.82 MiB (0.00%)

michaelwoerister commented:
Yep, that looks like a real regression. I'll look into it.

michaelwoerister commented:
If I'm reading the cachegrind diff right, this is the culprit:
https://github.com/rust-lang/rust/pull/119192/commits/62fc1e70cb2b8978171e242ce605be2666cdf5e7#diff-260026ed194b0aa8dd66fc4216bab5178f815f526842069ffc51a39708e7341fR1031

<details><summary>Details</summary>
<p>

```
--------------------------------------------------------------------------------
-- Summary
--------------------------------------------------------------------------------
Ir_______

4,578,847  PROGRAM TOTALS

--------------------------------------------------------------------------------
-- File:function summary
--------------------------------------------------------------------------------
  Ir________  file:function

<  4,610,220  ???:
   3,748,977    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
  -3,748,531    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
   2,871,292    core::slice::sort::recurse::<(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)), <[(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource))]>::sort_unstable_by<<rustc_data_structures::unord::UnordMap<rustc_lint_defs::LintId, (rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)>>::to_sorted_stable_ord::{closure#0}>::{closure#0}>
   1,958,585    <rustc_lint::levels::LintLevelsBuilder<rustc_lint::levels::TopDown>>::push
    -967,971    <rustc_parse::parser::Parser>::collect_tokens_for_expr::<<rustc_parse::parser::Parser>::parse_expr_prefix::{closure#7}>::{closure#0}
     907,050    <rustc_parse::parser::Parser>::parse_expr_prefix::{closure#7}
    -157,234    <rustc_parse::parser::Parser>::parse_expr_assoc_with
      60,534    rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
     -58,346    rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_hir::hir_id::OwnerId, rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>
      28,895    <rustc_middle::ty::Ty as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode
      17,188    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
     -17,186    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      12,963    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
     -12,942    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
     -11,715    <rustc_query_system::dep_graph::graph::DepGraphData<rustc_middle::dep_graph::DepsType>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>
     -11,595    <rustc_query_system::ich::hcx::StableHashingContext as rustc_span::HashStableContext>::def_path_hash
      11,589    <rustc_query_system::ich::hcx::StableHashingContext>::def_path_hash
     -11,198    <rustc_lint_defs::LintBuffer>::take
     -11,146    <rustc_lint::early::EarlyContextAndPass<rustc_lint::BuiltinCombinedPreExpansionLintPass> as rustc_ast::visit::Visitor>::visit_path
      11,029    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
     -11,028    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_resolve::Module)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_resolve::Module, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      10,961    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      -9,214    <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
       6,785    <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      -6,785    <hashbrown::raw::RawTable<(rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
      -5,364    <rustc_lint::BuiltinCombinedPreExpansionLintPass as rustc_lint::passes::EarlyLintPass>::check_ident
       5,286    core::slice::sort::insertion_sort_shift_left::<(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)), <[(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource))]>::sort_unstable_by<<rustc_data_structures::unord::UnordMap<rustc_lint_defs::LintId, (rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)>>::to_sorted_stable_ord::{closure#0}>::{closure#0}>

<    -23,951  /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/../multiarch/strcmp-sse2.S:strcmp

<    -14,678  <all-jemalloc-files>:<all-jemalloc-functions>

<      8,718  /usr/src/debug/glibc-2.36-18.fc37.x86_64/elf/../sysdeps/generic/dl-new-hash.h:_dl_lookup_symbol_x

<      4,963  /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/memcmp-avx2-movbe.S:__memcmp_avx2_movbe

--------------------------------------------------------------------------------
-- Function:file summary
--------------------------------------------------------------------------------
  Ir________  function:file

>  3,748,977  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

> -3,748,531  <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::hir_id::ItemLocalId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>  2,871,292  core::slice::sort::recurse::<(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)), <[(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource))]>::sort_unstable_by<<rustc_data_structures::unord::UnordMap<rustc_lint_defs::LintId, (rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)>>::to_sorted_stable_ord::{closure#0}>::{closure#0}>:???

>  1,958,585  <rustc_lint::levels::LintLevelsBuilder<rustc_lint::levels::TopDown>>::push:???

>   -967,971  <rustc_parse::parser::Parser>::collect_tokens_for_expr::<<rustc_parse::parser::Parser>::parse_expr_prefix::{closure#7}>::{closure#0}:???

>    907,050  <rustc_parse::parser::Parser>::parse_expr_prefix::{closure#7}:???

>   -157,234  <rustc_parse::parser::Parser>::parse_expr_assoc_with:???

>     60,534  rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>:???

>    -58,346  rustc_query_system::query::plumbing::ensure_must_run::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_hir::hir_id::OwnerId, rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt>:???

>     28,895  <rustc_middle::ty::Ty as rustc_serialize::serialize::Decodable<rustc_middle::query::on_disk_cache::CacheDecoder>>::decode:???

>    -23,951  strcmp:/usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/../multiarch/strcmp-sse2.S

>     17,188  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_span::def_id::CrateNum, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>    -17,186  <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_ast::node_id::NodeId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>    -14,678  <all-jemalloc-functions>:<all-jemalloc-files>

>     12,963  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<rustc_span::def_id::DefId>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>    -12,942  <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_hir::def::PartialRes, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>    -11,715  <rustc_query_system::dep_graph::graph::DepGraphData<rustc_middle::dep_graph::DepsType>>::try_mark_previous_green::<rustc_query_impl::plumbing::QueryCtxt>:???

>    -11,595  <rustc_query_system::ich::hcx::StableHashingContext as rustc_span::HashStableContext>::def_path_hash:???

>     11,589  <rustc_query_system::ich::hcx::StableHashingContext>::def_path_hash:???

>    -11,198  <rustc_lint_defs::LintBuffer>::take:???

>    -11,146  <rustc_lint::early::EarlyContextAndPass<rustc_lint::BuiltinCombinedPreExpansionLintPass> as rustc_ast::visit::Visitor>::visit_path:???

>     11,029  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, rustc_resolve::late::LifetimeUseSet, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>    -11,028  <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, rustc_resolve::Module)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, rustc_resolve::Module, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>     10,961  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>     -9,214  <hashbrown::raw::RawTable<(rustc_ast::node_id::NodeId, ())>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_ast::node_id::NodeId, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>      8,991  _dl_lookup_symbol_x:
       8,718    /usr/src/debug/glibc-2.36-18.fc37.x86_64/elf/../sysdeps/generic/dl-new-hash.h

>      6,785  <hashbrown::raw::RawTable<(rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_span::def_id::LocalDefId, alloc::vec::Vec<(rustc_middle::hir::place::Place, rustc_middle::mir::syntax::FakeReadCause, rustc_hir::hir_id::HirId)>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>     -6,785  <hashbrown::raw::RawTable<(rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>)>>::reserve_rehash::<hashbrown::map::make_hasher<rustc_hir::hir_id::ItemLocalId, alloc::vec::Vec<rustc_middle::ty::Ty>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>:???

>     -5,364  <rustc_lint::BuiltinCombinedPreExpansionLintPass as rustc_lint::passes::EarlyLintPass>::check_ident:???

>      5,286  core::slice::sort::insertion_sort_shift_left::<(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)), <[(&rustc_lint_defs::LintId, &(rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource))]>::sort_unstable_by<<rustc_data_structures::unord::UnordMap<rustc_lint_defs::LintId, (rustc_lint_defs::Level, rustc_middle::lint::LintLevelSource)>>::to_sorted_stable_ord::{closure#0}>::{closure#0}>:???

>      4,963  __memcmp_avx2_movbe:/usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/memcmp-avx2-movbe.S

--------------------------------------------------------------------------------
-- Annotated source file: /usr/src/debug/glibc-2.36-18.fc37.x86_64/elf/../sysdeps/generic/dl-new-hash.h
--------------------------------------------------------------------------------
Unannotated because one or more of these original files are unreadable:
- /usr/src/debug/glibc-2.36-18.fc37.x86_64/elf/../sysdeps/generic/dl-new-hash.h

--------------------------------------------------------------------------------
-- Annotated source file: /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/../multiarch/strcmp-sse2.S
--------------------------------------------------------------------------------
Unannotated because one or more of these original files are unreadable:
- /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/../multiarch/strcmp-sse2.S

--------------------------------------------------------------------------------
-- Annotated source file: /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/memcmp-avx2-movbe.S
--------------------------------------------------------------------------------
Unannotated because one or more of these original files are unreadable:
- /usr/src/debug/glibc-2.36-18.fc37.x86_64/string/../sysdeps/x86_64/multiarch/memcmp-avx2-movbe.S

--------------------------------------------------------------------------------
-- Annotated source file: <all-jemalloc-files>
--------------------------------------------------------------------------------
Unannotated because one or more of these original files are unreadable:
- <all-jemalloc-files>

--------------------------------------------------------------------------------
-- Annotation summary
--------------------------------------------------------------------------------
Ir_______

        0    annotated: files known & above threshold & readable, line numbers known
        0    annotated: files known & above threshold & readable, line numbers unknown
        0  unannotated: files known & above threshold & two or more non-identical
  -24,948  unannotated: files known & above threshold & unreadable
   -6,425  unannotated: files known & below threshold
4,610,220  unannotated: files unknown
```

</p>
</details>

michaelwoerister committed:
Provide generalized collect methods for UnordItems

michaelwoerister commented:
Let's see if https://github.com/rust-lang/rust/pull/119192/commits/1530b8b56529695997e06f33566a10a17c5bbfe6 does the trick.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1530b8b56529695997e06f33566a10a17c5bbfe6 with merge 198f5d0267d73f0c061ae1297dfda30c2bdd1f9f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1530b8b56529695997e06f33566a10a17c5bbfe6","merge_sha":"198f5d0267d73f0c061ae1297dfda30c2bdd1f9f"} -->

michaelwoerister commented:
@rust-lang/wg-compiler-performance, the way rustc-perf makes it easy to run cachegrind diffs is incredibly useful, by the way. I really appreciate all the effort that is going into this infrastructure!

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7298621311/job/19892069484)
Build commit: 198f5d0267d73f0c061ae1297dfda30c2bdd1f9f (`198f5d0267d73f0c061ae1297dfda30c2bdd1f9f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7298621311/job/19892069484"},"merge_sha":"198f5d0267d73f0c061ae1297dfda30c2bdd1f9f"} -->

rust-timer commented:
Queued 198f5d0267d73f0c061ae1297dfda30c2bdd1f9f with parent ef1b78eabe713a2068a1b0451102853dd2475a7b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (198f5d0267d73f0c061ae1297dfda30c2bdd1f9f): [comparison URL](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.1%, 4.5%] | 162   |
| Regressions ❌ <br /> (secondary)  | 1.6% | [0.2%, 6.4%] | 97    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.1%, 4.5%] | 162   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6% | [0.7%, 3.7%] | 6     |
| Regressions ❌ <br /> (secondary)  | 1.3% | [0.5%, 2.5%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.6% | [0.7%, 3.7%] | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.9% | [2.4%, 3.3%] | 5     |
| Regressions ❌ <br /> (secondary)  | 4.1% | [2.5%, 5.5%] | 7     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.9% | [2.4%, 3.3%] | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=198f5d0267d73f0c061ae1297dfda30c2bdd1f9f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.496s -> 673.938s (0.07%)
**Artifact size**: 312.79 MiB -> 312.79 MiB (-0.00%)

michaelwoerister commented:
That is not what I expected 😅

michaelwoerister commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit e71af937fec12a83fbceeba06439a3d6f740a774 with merge e0bf0d3f6472a2d92348bc5a990fa535f7461e2f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e71af937fec12a83fbceeba06439a3d6f740a774","merge_sha":"e0bf0d3f6472a2d92348bc5a990fa535f7461e2f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7300566121/job/19897735715)
Build commit: e0bf0d3f6472a2d92348bc5a990fa535f7461e2f (`e0bf0d3f6472a2d92348bc5a990fa535f7461e2f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7300566121/job/19897735715"},"merge_sha":"e0bf0d3f6472a2d92348bc5a990fa535f7461e2f"} -->

rust-timer commented:
Finished benchmarking commit (e0bf0d3f6472a2d92348bc5a990fa535f7461e2f): [comparison URL](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=e0bf0d3f6472a2d92348bc5a990fa535f7461e2f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=e0bf0d3f6472a2d92348bc5a990fa535f7461e2f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.1%, 0.4%]   | 9     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.3%, -0.2%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-0.3%, 0.2%]  | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=e0bf0d3f6472a2d92348bc5a990fa535f7461e2f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4% | [0.7%, 2.7%] | 5     |
| Regressions ❌ <br /> (secondary)  | 1.4% | [0.5%, 2.9%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.4% | [0.7%, 2.7%] | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=e0bf0d3f6472a2d92348bc5a990fa535f7461e2f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [2.0%, 2.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ef1b78eabe713a2068a1b0451102853dd2475a7b&end=e0bf0d3f6472a2d92348bc5a990fa535f7461e2f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.496s -> 673.262s (-0.03%)
**Artifact size**: 312.79 MiB -> 312.78 MiB (-0.00%)

michaelwoerister commented:
OK, that looks better. The fact that [this change](https://github.com/rust-lang/rust/commit/62fc1e70cb2b8978171e242ce605be2666cdf5e7#diff-260026ed194b0aa8dd66fc4216bab5178f815f526842069ffc51a39708e7341fR1031) caused such a big regression might suggest that there is some super-linear complexity in this code.

The remaining regressions seem to be in the `resolve_bound_vars` query. A fix there would probably be more work, because I just replace `FxHashMap` with `FxIndexMap` there.

TaKO8Ki commented:
@rustbot author

michaelwoerister commented:
Thanks for the review, @cjgillot! I'll follow up asap.

michaelwoerister commented on compiler/rustc_middle/src/middle/region.rs:
```diff
@@ -306,12 +307,12 @@ pub struct ScopeTree {
     /// Used to sanity check visit_expr/visit_pat call count when
     /// calculating coroutine interiors.
-    pub body_expr_count: FxHashMap<hir::BodyId, usize>,
+    pub body_expr_count: UnordMap<hir::BodyId, usize>,
```
Indeed, this disappears during rebasing.

michaelwoerister committed:
Address review comments and add back some #[inline] attrs from removed commits.

michaelwoerister commented:
I removed the two commits touching the LintLevelBuilder and addressed the comments (with the `body_expr_count` field gone, we can use `#[derive(HashStable)]` instead of the manual impl for `ScopeTree` \o/)

I don't think any of these changes warrant a new perf run.

@bors r=cjgillot

bors commented:
:pushpin: Commit 077540cedf836b3c6e17db39de21e2278bae90fc has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 077540cedf836b3c6e17db39de21e2278bae90fc -->
<!-- homu: {"type":"Approved","sha":"077540cedf836b3c6e17db39de21e2278bae90fc","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 077540cedf836b3c6e17db39de21e2278bae90fc with merge b8c207435c85955d0c0806240d5491f5ccd1def5...
<!-- homu: {"type":"BuildStarted","head_sha":"077540cedf836b3c6e17db39de21e2278bae90fc","merge_sha":"b8c207435c85955d0c0806240d5491f5ccd1def5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7426007723/job/20211809524)
Approved by: cjgillot
Pushing b8c207435c85955d0c0806240d5491f5ccd1def5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7426007723/job/20211809524"},"merge_sha":"b8c207435c85955d0c0806240d5491f5ccd1def5"} -->

rust-timer commented:
Finished benchmarking commit (b8c207435c85955d0c0806240d5491f5ccd1def5): [comparison URL](https://perf.rust-lang.org/compare.html?start=11035f9f52c65739d9613f5b9ecd332e9e640e71&end=b8c207435c85955d0c0806240d5491f5ccd1def5&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=11035f9f52c65739d9613f5b9ecd332e9e640e71&end=b8c207435c85955d0c0806240d5491f5ccd1def5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.1%, 0.5%]   | 7     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.6%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-0.6%, -0.2%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=11035f9f52c65739d9613f5b9ecd332e9e640e71&end=b8c207435c85955d0c0806240d5491f5ccd1def5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [0.9%, 1.5%]   | 8     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-3.7%, -0.8%] | 7     |
| All ❌✅ (primary)                 | 1.3%  | [0.9%, 1.5%]   | 8     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=11035f9f52c65739d9613f5b9ecd332e9e640e71&end=b8c207435c85955d0c0806240d5491f5ccd1def5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.9% | [-4.8%, -2.9%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=11035f9f52c65739d9613f5b9ecd332e9e640e71&end=b8c207435c85955d0c0806240d5491f5ccd1def5&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 668.063s -> 665.928s (-0.32%)
**Artifact size**: 311.09 MiB -> 311.13 MiB (0.01%)
